Add tests for update-alternatives and expand selection of tested shells

anonyco opened the pull request:
## Description

I'm currently learning CI/CD, so this is great practice for me.

## How Has This Been Tested?

This proposal doesn't add features and didn't change one byte in the index.sh. This is solely for improving and expanding the tests.

## Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] If I added new user-facing functionality, I have made corresponding changes to the README documenting the changes
- [x] If this is a code change, I have updated the [test configuration](https://github.com/benweissmann/getmic.ro/blob/master/.github/workflows/test.yml) and/or [test scripts](https://github.com/benweissmann/getmic.ro/tree/master/ci) to cover the changes I've made.

anonyco committed:
Expanded selection of tested shells, added tests for update-alternatives, and added README checksum tests

benweissmann commented on .github/workflows/test.yml:
```diff
@@ -94,22 +107,54 @@ jobs:
+        - 'busybox sh'
+        - gash
+        - mrsh
+        #- 'toybox toysh'
+        #- oil
```
What's up with these? Did they not work and we should delete them?

anonyco replied:
These are completely defunct shells that have such serious issues with basic POSIX syntax (e.x. toybox doesn't support `||` or multiline strings and oil has issues with variables requiring a special oil-only syntax in order to be assigned) that supporting these niche shells would be practically impossible.

Nonetheless, as these projects mature over time, it is possible that they may eventually gain the needed posix syntax in order to be worthy of consideration, so I'm leaving our options open for the future. Both these projects are far further along towards POSIX compliance than they were a few years ago.

So, let's not delete these and keep them commented out. The comment will serve as a reminder to check back on the status of the projects in a few years.

benweissmann commented on .github/workflows/test.yml:
```diff
@@ -94,22 +107,54 @@ jobs:
+        GETMICRO_REGISTER: ${{ matrix.update_alternatives }}
       run: ./ci/runDockerTest.sh alpine:latest sh
+
+  check_readme_checksum:
```
The checksum gets automatically updated on publish (you can check out the publishing script here: https://github.com/benweissmann/getmic.ro/blob/master/publish.sh) -- I think it's important that it gets updated at publish time, and not when a change merges, so the README isn't outdated in between when the PR is merged and when I publish the new version. So I think this is a cool idea but not a good fit for the CI pipeline.

anonyco replied:
Understood. I am removing it.

This line of code really rubs me the wrong way, so I'm going to do a little fix up if you don't mind, ha ha.

```sh
OLDSHA=$(rg -o 'The sha256 checksum is `([a-f0-9]+)`' -r '$1' < README.md)
```

benweissmann reviewed (changes requested):
This looks great. Added a few notes to the PR.

I think we should add one additional test (doesn't neccesarily need to run the matrix, could just be a one-off addition) to test the interactive prompt. I think you could use something like [empty](http://empty.sourceforge.net/) or a combination of [socat](https://linux.die.net/man/1/socat) and [yes](https://linux.die.net/man/1/yes) to simulate a user getting the prompt and entering yes or no.

anonyco committed:
Remove checksum check of README from CI

anonyco commented:
Also, one quick unrelated question: I am curious why you are using Amazon AWS to host the website instead of GitHub pages (where you could just copy index.sh to a subfolder named ghpages and rename it to index.html).

anonyco committed:
Revamped index.sh to my liking

anonyco committed:
Store shasum as comment in the markdown file

anonyco committed:
Fix for Issue #33

This should fix #33

anonyco committed:
Extra fix for #33 and https://github.com/zyedidia/micro/issues/2349

Hopefully, this will close the case shut once and for all.

Yes, I recognize that these notice messages are excessive and I know you are tempted to get rid of this feature, but we can't let the small portion of Linux users who can't read a basic straightforward error message get in the way of helping so many people who love this feature.

benweissmann commented on index.sh:
```diff
@@ -298,10 +309,11 @@ EOM
       cat 1>&2 << 'EOM'
 /=====================================\\
-|       INSUFFICIENT PRIVILEGES       |
+|    PLEASE READ THIS ERROR MESSAGE    |
```
Couple ideas here:
- Could we just avoid prompting the user if they don't have permission to write to `/etc/alternatives`? E.g. if they don't have write permission:
  -  if `GETMICRO_REGISTER` is undefined, print a brief warning letting them know that `update-alternatives` support is available but they would need to have root privileges (so the next time they use `getmicro`, they can use `sudo` and get this feature
  - If `GETMICRO_REGISTER=y`, abort with this error message
  - If `GETMICRO_REGISTER=n`, don't print any error or warning

benweissmann replied:
Another option: could we detect whether the system supports `sudo` and if so offer to just elevate permissions?

E.g. `Insufficient permissions to use update-alternatives. Try again with sudo? [y/N]: ` and then just re-run the block above with sudo if they select `y`.

anonyco replied:
I think I have the solution: the error message gives you the option to abort trying to use update-alternatives.

We can't automatically run sudo because then people like mrx23dot will install micro into their home directory, which will cause all sorts of issues when another user tries to use a system tool dependent on editor and can't run editor due to lack of permissions.

anonyco replied:
I love your solution to inform people about the GETMICRO_REGISTER but not prompt them by default.

This way, people who read will benefit from this feature and people who don't won't bother you, which is the best of both worlds. Additionally, anyone annoyed by the constant prompting for this will be relieved by the removal of the prompt.

The only issue with this is that it severely detracts from ease of use. The majority of people installing micro likely want this feature, so we shouldn't penalize them for this. 

Thus, I propose that we add a https://getmic.ro/r script. No extra units tests needed as this script is the same as index.sh except that it has an added `GETMICRO_REGISTER=y` at the start and is auto-generated.

Your other solution--to run sudo--would turn out quite problematic for people like mrx23dot. We can't automatically run sudo because people will use update-alternatives and install micro into their home directory, which will cause all sorts of issues when another user tries to use a system tool dependent on editor and can't run editor due to lack of permissions.

anonyco replied:
I just realized another issue with prompting for sudo and why making this an opt-in no-prompts feature really is the best way to go: multi-lingual speakers who struggle with English.

benweissmann commented on index.sh:
```diff
@@ -243,18 +243,29 @@ if [ "${altcmd:-x}" != "x" ] ; then
+(IMPORTANT: Choose n if you are busy, confused, or might open a GitHub Issue.)
 EOM
-    cpt="Register '$wrkdir/micro' with update-alternatives [y/n]: "
+    cpt="Register '$wrkdir/micro' with update-alternatives (prefer n if busy) [y/n]: "
```
I think we can just use the standard shell UX pattern for a default of no, e.g: `Register '$wrkdir/micro' with update-alternatives (prefer n if unsure) [y/N]: ` (note the capital `N`)

anonyco replied:
I was unaware of that standard UX pattern, but that makes a lot of sense. Thank you for bringing this to my attention.

I am making this change.

benweissmann commented on index.sh:
```diff
@@ -243,18 +243,29 @@ if [ "${altcmd:-x}" != "x" ] ; then
 Many people find it useful to make micro available on the PATH. One way to do
  this is to enter a root shell and run `cd /usr/bin` prior to installation.
 
+(IMPORTANT: Choose n if you are busy, confused, or might open a GitHub Issue.)
```
I think this is unnecessarily aggressive; `prefer n if unsure` is fine.

anonyco replied:
I agree. It is much too excessive. I'm not always the best at dealing with other people; it's something I need to work on.

Thank you.

benweissmann commented on publish.sh:
```diff
@@ -2,8 +2,34 @@
+  # according to order of operations, this is ! ( cmdexists aws && cmdexists git )
+  if ! cmdexists aws && cmdexists git ; then
+    echo 'Need to install aws and git in order to proceed' 1>&2
+    sudo apt-get install -y awscli git
```
I'm not a fan of installing stuff automatically here; this is an administrative script that's only ever run by maintainers so I don't think we need this check here -- if you have write permission to the production infrastructure, I think we can assume you know what to do if you get the error `aws: no such command`.

Definitely appreciate the cross-platform support and the better technique for getting `OLDSHA`!

anonyco replied:
Understood.

benweissmann commented on index.sh:
```diff
@@ -243,18 +243,29 @@ if [ "${altcmd:-x}" != "x" ] ; then
 |     update-alternatives detected     |
 \\=====================================/
 
+(Notice: getmicro must be run as root, e.x.: curl https://getmic.ro | sudo sh
```
As discussed below, I think we can just improve how we handle non-root users here; hopefully that makes this extra warning unneccesary.

anonyco replied:
I agree. This message is no longer necessary.

benweissmann commented on index.sh:
```diff
@@ -243,18 +243,29 @@ if [ "${altcmd:-x}" != "x" ] ; then
+(Notice: getmicro must be run as root, e.x.: curl https://getmic.ro | sudo sh
+ otherwise, trying to y use this feature will fail.)
+ 
+(Notice: if you y use this feature and you put micro in a location unaccessible
```
Can we combine this with the note suggesting putting `micro` in `/usr/bin` below?

anonyco replied:
Good idea. I will do this.

benweissmann reviewed (changes requested):
re: using AWS, that's a great question! It has a few advantages:

- We can serve the root object as `Content-Type: text/plain` so when you go to https://getmic.ro in a browser, it renders correctly (using Github Pages and having this file be `index.html`, we'd end up serving it as `text/html` which would make it render funny in browsers)
- AWS has more sophisticated security tools -- given that we're suggesting piping this script into `bash` (potentially as root!), it's obviously very important that we minimize the risk of an adversary being able to put a malicious script up. Serving the file from AWS gives us a couple advantages on that front:
  - It allows us to separate "can publish to getmic.ro" from "can push to / administer the repo"
  - AWS allows me to require hard-key 2FA (e.g. a yubikey) rather than SMS or app-based 2fa
  - The s3 bucket is set up with versioning, MFA delete, and tamper-proof Cloudtrail audit logs so we get an immutable record of any changes made to the file
- AWS allows me to keep basic access logs (e.g. IP and access time) when you use getmic.ro, which is useful for getting general usage stats (e.g. how many people use getmic.ro, what countries they're from for i18n prioritization purposes, etc), and could also be useful to contact impacted users in the event of a compromise.

mrx23dot commented:
From other issue:

.. it doesn't work out of the box with the current instructions (gives error),
nor does apt install the latest version.
So entry of barrier is put really high.

We could suggest in readme:
`curl https://getmic.ro | sh && sudo mv micro /usr/bin`
that only gives sudo to mv cmd, but **.sh script needs to be cleaned up, so it should end with ./micro available.**

Also I would like to get it working with ansible automation, for that I would **need silent .sh argument.** (to prevent asking for user input)

benweissmann commented:
- The messaging in the script already suggests moving the binary to `/usr/bin`; i'm not sure we need to put that in the README (there's many different "correct" locations depending on your system! `/usr/bin` is not appropriate in many cases).

- You can use [`GETMICRO_REGISTER`](https://github.com/benweissmann/getmic.ro#advanced-usage) to control this feature from non-interactive environments.

anonyco commented:
Hi @mrx23dot

Thank you for bringing this to my attention. I am working on a fix that makes the relevant information more obvious so that it's less confusing.

To answer your questions, I am expanding the help messages to encompass much more detail.

For automation, getmicro will not show any prompt if no interactive TTY is detected (so it will work just fine with Ansible) and default to N.

For your current situation, just choose N when you install micro and it will work alright.

Also, thank you for bringing ` curl https://getmic.ro | sh && sudo mv micro /usr/bin` to my attention. I did not consider that people would try to move micro after downloading it, so I will add an additional notice about this. (Using the update-alternatives feature and then moving micro will render /usr/bin/editor nonfunctional and probably break a lot of system tools like crontab).

anonyco commented:
@benweissmann We can't put that in the readme. Using update-alternatives in conjunction with `curl https://getmic.ro | sh && sudo mv micro /usr/bin` will break many system tools like crontab because /usr/bin/editor will be pointing to the prior location of micro. I'm adding a warning about this.

anonyco commented:
> re: using AWS, that's a great question! It has a few advantages:

Thank you for this explanation. That makes a lot of sense and helps broaden my knowledge of things.

anonyco committed:
Exclude the generated r file from the repo

anonyco committed:
Copypasta and 

* Made it easy to copypasta the awscli and git commands if apt-get is your cup of tea.

* Added https://getmic.ro/r script to automatically set GETMICRO_REGISTER=y

anonyco committed:
No more prompting the user to register with update-alternatives

Resolves: https://github.com/benweissmann/getmic.ro/pull/32#discussion_r800811298

anonyco committed:
Fixed lint warnings

anonyco committed:
Fixed test to correctly fail when micro fails to install

anonyco commented:
> This looks great. Added a few notes to the PR.

This is no longer needed as there is no interactive prompt anymore

anonyco committed:
Fix for mrsh

See https://github.com/benweissmann/getmic.ro/runs/5098989286?check_suite_focus=true

anonyco commented:
@benweissmann Apparently, we are running into a 429 rate limit error with vagrant, so try re-running the tests another time and they should all pass.

anonyco commented:
vagrant appears to be returning 429 rate limit exceeded errors, which is odd. One would think that GitHub would have an in-house proxy server to cache HTTP requests, but that is apparently not the case. 

@benweissmann Try re-running the tests and they should all pass.

benweissmann commented on index.sh:
```diff
@@ -240,21 +240,37 @@ if [ "${altcmd:-x}" != "x" ] ; then
 For example, this will allow `crontab -e` open the cron file with micro.
 
-To avoid this prompt in the future, define the GETMICRO_REGISTER variable. E.x:
+To enable this opt-in feature, define the GETMICRO_REGISTER variable. E.x:
```
Let's simplify this and just point people to `https://getmic.ro/r` (we can note `GETMICRO_REGISTER` in the README, but i'm not sure it's helpful to give 3 options here).

benweissmann replied:
Emphasizing `getmic.ro/r` has another small benefit, which is that I can use the server access logs to get a sense of what percentage of folks use this feature.

anonyco replied:
This is done.

I'm leaving one plain alternative without getmic.ro/r because it helps to know what is going on when you're trying to automate the installation. It might be much easier to have a switch in your auto install script for a environment variable than for a URL.

Also, good idea about usage statistics, however, getmic.ro/r will always be an under representation and there's no way to know how severely it underrepresents.

That is, some people will still be using plain old getmic.ro with GETMICRO_REGISTER=y

With that pessimism out of the way, I doubt that the underrepresentation will be that severe. The entire purpose of getmic.ro is to type less characters into the terminal and getmic.ro/r greatly reduces the amount you need to type, so I doubt many people will be using GETMICRO_REGISTER=y with plain old getmic.ro. However, automated installer scripts are another thing, where it could possibly go either way (if it were me writing the automated installer script, I would personally prefer the longer, more self-descriptive, version so I don't need to add a comment explaining it).

benweissmann commented on index.sh:
```diff
@@ -240,21 +240,37 @@ if [ "${altcmd:-x}" != "x" ] ; then
 EOM
-    cpt="Register '$wrkdir/micro' with update-alternatives [y/n]: "
+    doRegister="n"
+    cat >/dev/null << 'NUL'
```
I like the idea of skipping the prompt entirely and just pointing folks to getmic.ro/r -- let's go ahead and skip this prompt.

anonyco replied:
Yes. Sorry about the awkward multi-line string comment. It was necessary in order to fix mrsh: https://github.com/benweissmann/getmic.ro/runs/5098989286?check_suite_focus=true

I guess it's a good thing I added these extra shell script implementations, ha ha.

benweissmann commented on index.sh:
```diff
@@ -301,7 +319,8 @@ EOM
 \\=====================================/
 
-Uh oh! We couldn't run update-alternatives due to insufficient privileges.
+There is a very easy fix for this error, as explained below. We couldn't run
```
I think we can keep this error message as-is now that this block is opt-in.

anonyco replied:
I have done this.

benweissmann commented on index.sh:
```diff
@@ -240,21 +240,37 @@ if [ "${altcmd:-x}" != "x" ] ; then
     cat 1>&2 << 'EOM'
 /=====================================\\
-|     update-alternatives detected     |
+|   update-alternatives is supported   |
```
If `GETMICRO_REGISTER = y`, I think we should skip this message and instead just print a header indicating that we're going ahead with the `update-alternatives` code (I think it's confusing if you run getmic.ro, get this message, and then re-run getmic.ro/r and get this prompt again even though we're going to go ahead and do the registration)

anonyco replied:
I agree.

The if condition on lines 236 and 238 will skip showing the message if GETMICRO_REGISTER is already defined.

 https://github.com/benweissmann/getmic.ro/blob/13ca95d7d65b24aa9293b88fe080e275b7a0daea/index.sh#L236-L244

anonyco committed:
Cleanup TTY code, case insensitivity fix, diagnostic message, and trigger test run

anonyco committed:
Fix mrsh and fix permission errors

anonyco committed:
Fix lint warning and failed tests

In an epic moment of brainless unthinking, I wrote `n` on line 291, which inverted the results of whether micro was to be installed

anonyco committed:
Three times a charm?

Thank God for CI saving my butt.

anonyco commented on index.sh:
```diff
@@ -270,13 +281,22 @@ EOM
+      echo '/====================================='\\
+      echo '| Registering with update-alternatives |'
+      echo '\\=====================================/'
+      echo
```
Is this new status message okay?

Other than this, I think we're all good. All tests are passing. I'm ready to push this if you are.

benweissmann commented on index.sh:
```diff
@@ -226,35 +227,45 @@ elif command -v update-alternatives >/dev/null 2>&1 ; then
+    cat >/dev/null << 'NUL'
+    # Commenting this out resolves https://github.com/benweissmann/getmic.ro/pull/32#discussion_r800962643
+    
+    cpt="Register '$wrkdir/micro' with update-alternatives (prefer n if unsure) [y/N]: "
```
I though the idea here was that we wouldn't prompt unless the user is root / can run update-alternatives? If i'm reading this correctly, we still end up prompting non-root users here, no?

anonyco replied:
All of this is in a comment block starting on line 265. No user gets prompted for input ever.

All users, regardless of whether they are root or not, will receive the info blurb about update-alternatives if its available on their system.

benweissmann committed:
Remove commented code; massage the copy a little bit

benweissmann commented:
Apologies for the delay; it's been a busy week. I've pushed up a couple small changes:

- I removed rather than commented out the older prompt code; we can always grab it from git if needed and I'd like to keep the script as short as possible to be easier on folks who are reading it to figure out what it does before running it
- I made a few minor copy edits

Will merge as soon as tests finish running -- thanks again for your hard work on this!

benweissmann committed:
Fix typos in publish.sh

anonyco commented:
Thank you for keeping me on my toes and holding my feet to the fire. I learned a lot of new things through working on this, so I should be thanking you.
